vm/address/src/lib.rs

```diff
@@ -359,4 +359,92 @@ pub mod json {
+        use forest_json_utils::GoVecVisitor;
+        use serde::ser::SerializeSeq;
+
+        /// Wrapper for serializing and deserializing a Cid vector from JSON.
+        #[derive(Deserialize, Serialize)]
+        #[serde(transparent)]
+        pub struct AddressJsonVec(#[serde(with = "self")] pub Vec<Address>);
+
+        /// Wrapper for serializing a cid slice to JSON.
+        #[derive(Serialize)]
+        #[serde(transparent)]
+        pub struct AddressJsonSlice<'a>(#[serde(with = "self")] pub &'a [Address]);
+
+        pub fn serialize<S>(m: &[Address], serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            let mut seq = serializer.serialize_seq(Some(m.len()))?;
+            for e in m {
+                seq.serialize_element(&AddressJsonRef(e))?;
+            }
+            seq.end()
+        }
+
+        pub fn deserialize<'de, D>(deserializer: D) -> Result<Vec<Address>, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            deserializer.deserialize_any(GoVecVisitor::<Address, AddressJson>::new())
+        }
+    }
+
+    pub mod opt {
+        use super::*;
+        use serde::{self, Deserialize, Deserializer, Serializer};
+        use std::borrow::Cow;
+
+        pub fn serialize<S>(v: &Option<Address>, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            if let Some(unwrapped_address) = v.as_ref() {
+                serializer.serialize_str(&encode(unwrapped_address))
+            } else {
+                serializer.serialize_str("\u{003c}\u{0065}\u{006d}\u{0070}\u{0074}\u{0079}\u{003e}")
+            }
+        }
+
+        pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<Address>, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            let address_as_string: Cow<'de, str> = Deserialize::deserialize(deserializer)?;
+            if address_as_string == "\u{003c}\u{0065}\u{006d}\u{0070}\u{0074}\u{0079}\u{003e}" {
+                return Ok(None);
+            }
+            Ok(Some(
+                Address::from_str(&address_as_string).map_err(de::Error::custom)?,
+            ))
+        }
+    }
+
+    pub mod opt_vec {
+        use super::*;
+        use crate::json::AddressJsonRef;
+        use serde::ser::SerializeSeq;
+        use serde::{self, Serializer};
+
+        pub fn serialize<S>(v: &Option<Vec<Address>>, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            if let Some(unwrapped_addresses) = v.as_ref() {
+                let mut seq = serializer.serialize_seq(Some(unwrapped_addresses.len()))?;
+                for e in unwrapped_addresses {
+                    seq.serialize_element(&AddressJsonRef(e))?;
+                }
+                seq.end()
+            } else {
+                serializer.serialize_none()
+            }
+        }
+    }
```

Comment:
This should never be necessary, best not to fall into this pattern of wrapping empty vecs in an `Option`

Reply:
This is the difference between serialization of an empty vec that returns nil verses an empty vec that serializes an empty set. An empty optional vec will serialize as nil while an empty vec will serialize as [] https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ed7ce53123d1bc5418b3bb69bda0c8b2

this is matching the data packets that lotus is expecting, had it as a vector before but found failure

Reply:
I understand that is what Lotus provides to it's miner, my point is that the Lotus miner will parse the empty array the same as nil, and then we don't have to follow this bad pattern of ambiguous formats and handling the `None` case

Reply:
Okay,I had issues running an empty array but I will switch it to a vector

Reply:
I'm not saying that's absolutely the case, but there is other instances that had been tested and assume that the go implementation deserializes it as such. Are you sure, or can you reproduce this with this difference? I don't want to handle this pattern for every single vector just to match the go daemon implementation's json output, if it's functionally the same

Reply:
this can still be removed